util/dvsim/SimCfg.py

```diff
@@ -446,27 +443,49 @@ def _expand_run_list(self, build_map):
 
         return runs
 
+    def _is_unique_build(self, build_mode_obj):
+        '''Uniquifies the builds if a new one is identical to an existing one.
+
+        It is possible for tests to supply different build modes, but those
+        builds may differ only under specific circumstances, such as coverage
+        being enabled. If coverage is not enabled, then they may be completely
+        identical. In that case, we can save compute resources by removing the
+        extra duplicated builds.
+
+        The `build_mode_obj` arg is the BuildMode object created from HJson.
+        This method creates & returns the CompileSim object (if unique).
+        '''
+        new_build = CompileSim(build_mode_obj, self)
+        for build in self.builds:
+            if build.is_equal(new_build):
+                return (False, build)
+        return (True, new_build)
```

Comment:
This seems like it's conflating "make a Deploy object from `build_mode_obj`" and "check whether two `build_mode_obj` values match". Would it be possible to move the comparison that's currently in `Deploy.is_equal` to the build mode object? Then you could filter `self.build_list` and build `CompileSim` objects from what remained, getting rid of the slightly non-obvious tuple return value.

Reply:
Cant do that unfortunately because the key-value substitutions happen late (intentionally) as a part of `CompileSim` as opposed to in `BuildModes`. The `is_equal` check needs to be performed on the final, completely resolved `cmd` so that we know for sure two builds will behave identically. 

As an example, we set `default_vcs_cov_hier: {{build_mode}_vcs_cov_hier}`. This key is not a part of the `BuildModes` object; its a part of the parent `SimCfg` object. The [default, cover_reg_top] `BuildModes` objects will be identical, but the final `cmd` resolved after from them all the substitutions (in the `CompileSim` objects created from them) will be different because of the different coverage hierarchy files each build mode provides.

Reply:
I got rid of the `is_unique_build` and expanded the 4 lines inline where I was calling it. This is hopefully better.

Reply:
Ah, good point! Yep, that looks much cleaner.